### PR TITLE
Add SIGHASH_NONE, SIGHASH_SINGLE, SIGHASH_ANYONECANPAY hash types

### PR DESCRIPTION
**Pull request by dthorpe**
Test scripts written, but unable to run on Windows.

### CODE REVIEW

**Comment by dcousens** [comment]
Cheers @dthorpe, this is good :)

Could you be able to make it an extra test fixture rather than changing an existing? 

-- As for it not running on Windows, I haven't got a test suite to check it out, but I'll see what I can find out.

**Comment by dthorpe** [comment]
Sure, I can copy the signature tests out to a different fixture.

**Comment by dcousens** [comment]
Let me know when they've been added.

**Comment by dthorpe** [comment]
sighash test data pulled out to a new fixture

File: test/fixtures/transaction.json [hunk 1]
```diff
@@ -148,7 +148,7 @@
...
...ffdf9c7b01003060a6c841a01f4fff8a5a0010000008b483045022100991d1bf60c41358f08b20e53718a24e05ac0608915df4f6305a5b47cb61e5da7022003f14fc1cc5b737e2c3279a4f9be1852b49dbb3d9d6cc4c8af6a666f600dced8014104aa592c859fd00ed2a02609aad3a1bf72e0b42de67713e632c70a33cc488c15598a0fb419370a54d1c275b44380e8777fc01b6dc3cd43a416c6bab0e30dc1e19fffffffff4cba12549f1d70f8e60aea8b546c8357f7c099e7c7d9d8691d6ee16e7dfa3170010000008c493046022100f14e2b0ef8a8e206db350413d204bc0a5cd779e556b1191c2d30b5ec023cde6f022100b90b2d2bf256c98a88f7c3a653b93cec7d25bb6a517db9087d11dbd189e8851c014104aa592c859fd00ed2a02609aad3a1bf72e0b42de67713e632c70a33cc488c15598a0fb419370a54d1c275b44380e8777fc01b6dc3cd43a416c6bab0e30dc1e19fffffffffa4b3aed39eb2a1dc6eae4609d9909724e211c153927c230d02bd33add3026959010000008b483045022100a8cebb4f1c58f5ba1af91cb8bd4a2ed4e684e9605f5a9dc8b432ed00922d289d0220251145d2d56f06d936fd0c51fa884b4a6a5fafd0c3318f72fb05a5c9aa372195014104aa592c859fd00ed2a02609aad3a1bf72e0b42de67713e632c70a33cc488c15598a0fb419370a54d1c275b44380e8777fc01b6dc3cd43a416c6bab0e30dc1e19fffffffff0240d52303000000001976a914167c3e1f10cc3b691c73afbdb211e156e3e3f25c88ac15462e00000000001976a914290f7d617b75993e770e5606335fa0999a28d71388ac00000000"
```
Comment (dcousens) [inline test/fixtures/transaction.json]: Why is this still changing?

File: src/transaction.js [hunk 2]
```diff
@@ -200,18 +200,46 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
   txTmp.ins.forEach(function (txIn) {
     txIn.script = Script.EMPTY
   })
-  txTmp.ins[inIndex].script = hashScript
+  txTmp.ins[inIndex].script = hashScript;
```
Comment (dcousens) [inline src/transaction.js]: Please remove the semi-colons throughout, they break our test suite as we use [standard](https://github.com/feross/standard) on this project.

File: src/transaction.js [hunk 3]
```diff
@@ -200,18 +200,46 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
...
+    }
+    txTmp.outs.forEach(function (output, i) {
+      if (i != inIndex) {
+        txTmp.outs[i] = stubOut;
+      }
+    });
+
+    //All other inputs aside from the current input in txTmp have their nSequence index set to zero
+    txTmp.ins.forEach(function (input, i) {
+      if (i != inIndex) {
```
Comment (dcousens) [inline src/transaction.js]: `!==`

File: src/transaction.js [hunk 4]
```diff
@@ -200,18 +200,46 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
...
-    assert(false, 'SIGHASH_NONE not yet supported')
-  } else if (hashTypeModifier === Transaction.SIGHASH_SINGLE) {
-    assert(false, 'SIGHASH_SINGLE not yet supported')
+    //All other txTmp outputs aside from the output that is the same as the current input index are set to a blank script and a value of (long) -1.
+    var stubOut = {
+      script: Script.EMPTY,
+      value: -1
+    }
+    txTmp.outs.forEach(function (output, i) {
+      if (i != inIndex) {
```
Comment (dcousens) [inline src/transaction.js]: `!==`

File: src/transaction.js [hunk 5]
```diff
@@ -200,18 +200,46 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
...
+        txTmp.outs[i] = stubOut;
```
Comment (dcousens) [inline src/transaction.js]: Also,  maybe we should create an object each time.

``` javascript
txTmp.outs[i] = {
  script: Script.EMPTY,
  value: -1
}
```

File: src/transaction.js [hunk 6]
```diff
@@ -200,18 +200,46 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
...
+    txTmp.outs = txTmp.outs.slice(0, inIndex);
 
-  if (hashTypeModifier === Transaction.SIGHASH_NONE) {
-    assert(false, 'SIGHASH_NONE not yet supported')
-  } else if (hashTypeModifier === Transaction.SIGHASH_SINGLE) {
-    assert(false, 'SIGHASH_SINGLE not yet supported')
+    //All other txTmp outputs aside from the output that is the same as the current input index are set to a blank script and a value of (long) -1.
+    var stubOut = {
+      script: Script.EMPTY,
+      value: -1
```
Comment (dcousens) [inline src/transaction.js]: Is `-1`, equivalent to `0xffffffff` here?

**Comment by dcousens** [comment]
Please fix the formatting issues as described here: https://travis-ci.org/bitcoinjs/bitcoinjs-lib/jobs/59269303

File: test/fixtures/transaction.json (hunk 1 above)
Comment (dcousens) [inline test/fixtures/transaction.json]: Also, you added a superfluous comma here that is breaking the JSON.  Please remove it.

**Comment by dcousens** [comment]
I'm going to merge this into a separate branch, and work on it there.  Thanks for your work @dthorpe.

File: src/transaction.js [hunk 7]
```diff
@@ -200,18 +200,46 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
...
+    txTmp.outs.forEach(function (output, i) {
+      if (i != inIndex) {
+        txTmp.outs[i] = stubOut;
+      }
+    });
+
+    //All other inputs aside from the current input in txTmp have their nSequence index set to zero
+    txTmp.ins.forEach(function (input, i) {
+      if (i != inIndex) {
+        txCopy.Inputs[i].Sequence = 0;
```
Comment (dcousens) [inline src/transaction.js]: Also, `txCopy` was undefined.  Will have to fix this.

**Pull request closed without merging**